Added human player hopper subsystem

nickfrisch opened the pull request.

Commit:
Created BallHopper subsystem and began states

Commit:
Started to create custom init and run methods for slow release

Commit:
Began working on including BannerSensor

Commit:
Finished custom SlowReleaseState

witcpalek commented on src/main/config/robot.dtd:
```diff
@@ -129,7 +130,7 @@
           usage             	    ( DRIVE  | TURN  |
                                       INTAKE1 | INTAKE2 | BALL_TRANSFER | TURRET | 
-                                      SHOOTER_1 | SHOOTER_2 | SHOOTER_HOOD ) "DRIVE_MOTOR"
+                                      SHOOTER_1 | SHOOTER_2 | SHOOTER_HOOD | BALL_HOPPER_1 | BALL_HOPPER_2 ) "DRIVE_MOTOR"
```
Why are there 2 ball hoppers?

witcpalek commented on src/main/config/swerveBot/robot.xml:
```diff
@@ -259,6 +267,27 @@
                      reverselimitswitchopen="true"/>        
 	</mechanism>
+
+       <mechanism type="BALL_HOPPER">
```
Why are there 2 ball hoppers?   These CAN IDs are already used; need to pick unique IDs.

witcpalek commented on src/main/cpp/states/ballhopper/BallHopperSlowRelease.cpp:
```diff
@@ -0,0 +1,93 @@
+#include <hw/DragonDigitalInput.h>
+
+using namespace std;
+
```
This file needs some comments.    You're managing switching between sub-states, but without reading the entire file,  understanding whether you are dropping first or holding first as well as what triggers the switch between states isn't clear.   You don't want someone to come in thinking they are fixing an issue and break the whole scheme.

witcpalek commented on src/main/cpp/states/ballhopper/BallHopperSlowRelease.h:
```diff
@@ -0,0 +1,62 @@
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
+// OR OTHER DEALINGS IN THE SOFTWARE.
+//====================================================================================================================================================
+
```
Add doco

witcpalek commented on src/main/cpp/states/ballhopper/BallHopperSlowRelease.h:
```diff
@@ -0,0 +1,62 @@
+
+//Team 302 Includes
+#include <states/Mech1MotorState.h>
+#include <states/ballhopper/BallHopperStateMgr.h>
```
why do you need this include?

witcpalek commented on src/main/cpp/states/ballhopper/BallHopperSlowRelease.h:
```diff
@@ -0,0 +1,62 @@
+//Team 302 Includes
+#include <states/Mech1MotorState.h>
+#include <states/ballhopper/BallHopperStateMgr.h>
+#include <hw/DragonDigitalInput.h>
```
Why do you need this include?

witcpalek commented on src/main/cpp/states/ballhopper/BallHopperStateMgr.cpp:
```diff
@@ -0,0 +1,187 @@
+#include <xmlmechdata/StateDataDefn.h>
+#include <controllers/MechanismTargetData.h>
+#include <utils/Logger.h>
+#include <gamepad/TeleopControl.h>
```
why do you need this?

nickfrisch replied:
TargetData and StataDataDefn are used for parsing config data and I just forgot to remove TeleopControl
.

witcpalek commented on src/main/cpp/states/ballhopper/BallHopperStateMgr.cpp:
```diff
@@ -0,0 +1,187 @@
+
+                    case BALL_HOPPER_STATE::SLOW_RELEASE:
+                    {
+                        auto thisState = new BallHopperSlowRelease( controlData, target, GetState(BALL_HOPPER_STATE::HOLD), GetState(BALL_HOPPER_STATE::RAPID_RELEASE) );
```
Shouldn't do this.  It is assuming that these other states have already been created and they may not have been.   Instead this state should query the state manager for the states it needs when it initializes if it doesn't have them already.

witcpalek commented on src/main/cpp/states/ballhopper/BallHopperStateMgr.cpp:
```diff
@@ -0,0 +1,187 @@
+        {
+            SetCurrentState( BALL_HOPPER_STATE::OFF, false);
+        }
+        else if ( m_currentStateEnum == BALL_HOPPER_STATE::HOLD && m_currentStateEnum != BALL_HOPPER_STATE::HOLD)
```
Delete; this shouldn't be this complex just run the current state (if there is no current state, log an error)

witcpalek commented on src/main/cpp/states/ballhopper/BallHopperStateMgr.h:
```diff
@@ -0,0 +1,76 @@
+#pragma once
+
+//C++ Includes
+#include <map>
```
is this needed?

witcpalek commented on src/main/cpp/subsys/BallHopper.cpp:
```diff
@@ -0,0 +1,38 @@
+(
+    shared_ptr<IDragonMotorController>  motorController,
+    shared_ptr<DragonDigitalInput>      ballDetection
+) : Mech1IndMotor( MechanismTypes::MECHANISM_TYPE::BALL_HOPPER, string( "ballhopper.xml"), string("BallHopperNT"), motorController)
```
its preferable to do

) : Mech1IndMotor(....), m_bannerSensor(ballDetection)

Also, it would be better to not refer to it as a bannersensor rather what it is doing (if they change the sensor type for detecting balls, you don't want the variable to be misleading)

Commit:
Fixed commented areas

The creation of hold and release state in the slowreleasestate may need to be changed

Commit:
Fixed comments on pull request
